pyqtorch/primitive.py

```diff
@@ -0,0 +1,129 @@
+from __future__ import annotations
+
+from math import log2
+
+import torch
+
+from pyqtorch.abstract import AbstractOperator
+from pyqtorch.apply import apply_operator
+from pyqtorch.matrices import OPERATIONS_DICT, _dagger, make_controlled
+from pyqtorch.utils import Operator, State
+
+
+class Primitive(AbstractOperator):
+    def __init__(self, pauli: torch.Tensor, target: int):
+        super().__init__(target)
+        self.register_buffer("pauli", pauli)
+        self.qubit_support = [self.target]
+        self.n_qubits = len(self.qubit_support)
+
+    def unitary(self, values: dict[str, torch.Tensor] = {}) -> Operator:
+        return self.pauli.unsqueeze(2)
+
+    def forward(self, state: State, values: dict[str, torch.Tensor] = {}) -> State:
+        return apply_operator(
+            state, self.unitary(values), self.qubit_support, len(state.size()) - 1
+        )
+
+    def dagger(self, values: dict[str, torch.Tensor] = {}) -> Operator:
+        return _dagger(self.unitary(values))
+
+
+class X(Primitive):
+    def __init__(self, target: int):
+        super().__init__(OPERATIONS_DICT["X"], target)
```

Comment:
Just wondering if we can gain something in avoiding all these look-up overheads.

Reply:
thats done once in the initialization of the gate.

Reply:
What I meant is passing directly the matrix.

Reply:
dict lookup is O(1)

Reply:
what do you think @Roland-djee ?

Reply:
Yeah, I suppose for such a small finite dict it doesn't matter much.